src/Data/Array/Mutable/Unlifted/Linear.hs

```diff
@@ -0,0 +1,107 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE LinearTypes #-}
+{-# LANGUAGE MagicHash #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE UnboxedTuples #-}
+{-# LANGUAGE UnliftedNewtypes #-}
+
+-- |
+-- This module defines a linear wrapper around 'GHC.MutableArray#'
+-- which can be manipulated in pure contexts.
+--
+-- It ensures that effects happen in order. However, access to
+-- out-of-bounds indices causes undefined behaviour.
+--
+-- This module is meant to be imported qualified.
+module Data.Array.Mutable.Unlifted.Linear
+  ( Array#
+  , unArray#
+  , alloc
+  , allocBeside
+  , lseq
+  , size
+  , read
+  , write
+  , copyInto
+  ) where
+
+import Data.Unrestricted.Linear hiding (lseq)
+import Prelude (Int)
+import qualified Prelude as Prelude
+import qualified Unsafe.Linear as Unsafe
+import qualified GHC.Exts as GHC
+
+-- | A mutable array holding @a@s
+newtype Array# a = Array# (GHC.MutableArray# GHC.RealWorld a)
+
+-- | Run an unsafe computation using the underlying MutableArray,
+-- consuming the 'Array#' in process.
+unArray# :: (GHC.MutableArray# GHC.RealWorld a -> b) -> Array# a #-> Ur b
+unArray# f = Unsafe.toLinear (\(Array# a) -> Ur (f a))
+
+-- | Consume an 'Array#'.
+--
+-- Note that we can not implement a 'Consumable' instance because 'Array#'
+-- is unlifted.
+lseq :: Array# a #-> b #-> b
+lseq = Unsafe.toLinear2 (\_ b -> b)
+
+-- | Allocate a mutable array of given size using a default value.
+--
+-- The size should be non-negative.
+alloc :: Int -> a -> (Array# a #-> Ur b) #-> Ur b
+alloc (GHC.I# s) a f =
+  let !new = GHC.runRW# Prelude.$ \st ->
+        case GHC.newArray# s a st of
+          (# _, arr #) -> Array# arr
+   in f new
+
+-- | Allocate a mutable array of given size using a default value,
+-- using another 'Array#' as a uniqueness proof.
+--
+-- The size should be non-negative.
+allocBeside :: Int -> a -> Array# b #-> (# Array# b, Array# a #)
+allocBeside (GHC.I# s) a orig =
+  let !new = GHC.runRW# Prelude.$ \st ->
+        case GHC.newArray# s a st of
+          (# _, arr #) -> Array# arr
+   in (# orig, new #)
+
+size :: Array# a #-> (# Array# a, Ur Int #)
+size = Unsafe.toLinear go
+  where
+    go :: Array# a -> (# Array# a, Ur Int #)
+    go (Array# arr) =
+      let !s = GHC.sizeofMutableArray# arr
+      in  (# Array# arr, Ur (GHC.I# s) #)
+
+read ::  Int -> Array# a #-> (# Array# a, Ur a #)
+read (GHC.I# i) = Unsafe.toLinear go
+  where
+    go :: Array# a -> (# Array# a, Ur a #)
+    go (Array# arr) =
+      case GHC.runRW# (GHC.readArray# arr i) of
+        (# _, ret #) -> (# Array# arr, Ur ret #)
+
+write :: Int -> a -> Array# a #-> Array# a
+write (GHC.I# i) (a :: a) = Unsafe.toLinear go
+  where
+    go :: Array# a -> Array# a
+    go (Array# arr) =
+      case GHC.runRW# (GHC.writeArray# arr i a) of
+        _ -> Array# arr
```

Comment:
Does `Unsafe.toLinear` prevent this code from being inlined?
Otherwise, what stops
```Haskell
case GHC.runRW# (GHC.writeArray# arr i a) of
  _ -> case GHC.runRW# (GHC.writeArray# arr i b) of
    _ -> Array# arr
```
from being rewritten to
```Haskell
case GHC.runRW# (GHC.writeArray# arr i b) of
  _ -> case GHC.runRW# (GHC.writeArray# arr i a) of
    _ -> Array# arr
```

Reply:
I do not know; I don't think `Unsafe.toLinear` would prevent this code from being inlined, since it is nothing more than a `unsafeCoerce`. Do you think adding a bang to the wildcard pattern would fix it?

Reply:
I'm thinking `unsafeCoerce (\x -> f x) a` prevents GHC to simplify the expression to `f x` because it would be ill typed.
Even if `write` is inlined, I think this could somewhat prevent `go` from being affected by the surrounding code. But it is not a solid argument.

This discussion affects correctness of `read`, `write` and `copyInto`. One workaround would be to declare the three functions as non-inlineable.

Reply:
I will make those three functions as non-inlineable. I don't know how big of a performance hit it'll be, but correctness should come first.

Reply:
> I'm thinking `unsafeCoerce (\x -> f x) a` prevents GHC to simplify the expression to `f x` because it would be ill typed.

It doesn't, actually. There are special provisions in the compiler to let the computation go further (basically, this is changed to `(\x -> unsafeCoerce (f x)) (unsafeCoerce a)`. At present, `unsafe.toLinear` will, however, block this computation. But this is considered a limitation of the implementation, and will be fixed.

Generally, speaking, it is a good point that we ought to add `NOINLINE` on every function calling `runRW#`. Otherwise, issues like what @facundominguez pointed out may arise.

Reply:
> But this is considered a limitation of the implementation, and will be fixed.

It will be interesting to compare pure and monadic linear mutable arrays when this is fixed. This would make observable the cost introduced by NOINLINE.